include/triton/Tools/LinearLayout.h

```diff
@@ -0,0 +1,433 @@
+#ifndef TRITON_TOOLS_LINEARLAYOUT_H
+#define TRITON_TOOLS_LINEARLAYOUT_H
+
+#include <cstdint>
+#include <ostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/STLExtras.h"
+
+namespace mlir::triton {
+
+// # High-level overview of linear layouts
+//
+// The idea for linear layouts is due to Adam P. Goucher.
+//
+// In Triton, a linear layout (LL) is a function that maps from a "hardware
+// location" to a "logical tensor index".
+//
+// For example, suppose we have a 2D tensor T stored in GPU registers.  T's
+// layout is the function that, given a "hardware location" tuple of (thread-id,
+// warp-id), returns an index (x,y) into T.  In other words, if L(t,w) = (x,y)
+// is our linear layout func, then thread t in warp w contains the value T[x,y].
+//
+// The key fact about LLs is, the mapping from (t,w) to (x,y) is not arbitrary.
+// We only need to specify the value of L(t,w) at certain special points
+// (namely, the values L(t,0) and L(0,w) where t/w are powers of 2), and from
+// those we can compute all the other values of L.
+//
+// Here's an example LL where we have 4 warps and 4 threads per warp, and the
+// tensor T has shape 4x4.  We define the function L by choosing the values of
+// L(0,1), L(0,2), L(1,0), and L(2,0).  Our choices are shown below.
+//
+//               t/w   0     1     2    3
+//               0      ? (0,1) (0,2)   ?
+//    L(t,w) =   1  (1,1)     ?     ?   ?
+//               2  (2,2)     ?     ?   ?
+//               3      ?     ?     ?   ?
+//
+// You only need to specify these four values to define the whole linear layout.
+// These special values are called the "basis vectors" or "bases" of the layout.
+// We complete the table by xor'ing together the bases, according to the
+// following rule.  (I write "⊕" for xor.)
+//
+//    L(t1 ⊕ t2, w1 ⊕ w2) = L(t1, w1) ⊕ L(t2, w2)  (linearity rule).
+//
+// The linearity rule plus our four choices allows us to fill in the whole
+// table.  Here's how we might compute some of the values.
+//
+//    L(0,0) = L(1 ⊕ 1, 0 ⊕ 0) = L(1,0) ⊕ L(1,0) = (1,1) ⊕ (1,1) = (0,0)
+//    L(0,3) = L(0, 2 ⊕ 1)     = L(0,2) ⊕ L(0,1) = (0,2) ⊕ (0,1) = (0,3)
+//    L(3,0) = L(2 ⊕ 1, 0)     = L(2,0) ⊕ L(1,0) = (2,2) ⊕ (1,1) = (3,3)
+//    L(3,3) = L(3 ⊕ 0, 0 ⊕ 3) = L(3,0) ⊕ L(0,3) = (3,3) ⊕ (0,3) = (0,3).
```

Comment:
Thank you, will upload this in my next revision.

Since you read my Very Long Comment, is it understandable?  I've tried to provide lots of examples, but I'm also worried it's too long.  Any feedback on length or anything else?

Reply:
I like it! I really like the examples for how to construct the trivial layouts (identity, transpose, projection). It could be good to extend it with a section where you start with the usual formula for swizzling and you build the linear map from it. You can also leave a comment asking the reader to do so :)
I'm proposing this, because as soon as I saw this I was "ok, I guess these make sense to implement swizzling, but I'm not sure how yet..." :D